libraries/standard/mqtt/src/mqtt.c

```diff
@@ -375,41 +438,41 @@ MQTTStatus_t MQTT_Publish( MQTTContext_t * const pContext,
 
     if( status == MQTTSuccess )
     {
-        /* Send header first. */
-        bytesSent = sendPacket( pContext,
-                                pContext->networkBuffer.pBuffer,
-                                headerSize );
-
-        if( bytesSent < 0 )
+        /* Reserve state for publish message. Only to be done for QoS1 or QoS2. */
+        if( pPublishInfo->qos > MQTTQoS0 )
         {
-            LogError( "Transport send failed for PUBLISH header." );
-            status = MQTTSendFailed;
+            status = MQTT_ReserveState( pContext,
+                                        packetId,
+                                        pPublishInfo->qos );
         }
-        else
-        {
-            LogDebugWithArgs( "Sent %d bytes of PUBLISH header.",
-                              bytesSent );
+    }
+
+    if( status == MQTTSuccess )
+    {
+        /* Sends the serialized publish packet over network. */
+        status = sendPublish( pContext,
+                              pPublishInfo,
+                              headerSize );
+    }
 
-            /* Send Payload. */
-            bytesSent = sendPacket( pContext,
-                                    pPublishInfo->pPayload,
-                                    pPublishInfo->payloadLength );
-
-            if( bytesSent < 0 )
-            {
-                LogError( "Transport send failed for PUBLISH payload." );
-                status = MQTTSendFailed;
-            }
-            else
-            {
-                LogDebugWithArgs( "Sent %d bytes of PUBLISH payload.",
-                                  bytesSent );
-            }
+    if( status == MQTTSuccess )
```

Comment:
What happens if the publish was not successful? Should we add an `MQTT_PurgeState()` function to cleanup and remove the packet id from the state record?

Reply:
Should we add state only after publish is successful ?

Reply:
> What happens if the publish was not successful? Should we add an `MQTT_PurgeState()` function to cleanup and remove the packet id from the state record?

Yes I think, this function would be required. I will add  a TODO to address that in the comments once it is available.

Reply:
> Should we add state only after publish is successful ?

no. We should try to send a publish only after MQTT_ReserveState() is successful.